### PR TITLE
More elegant holders for local lazy vals.

### PR DESCRIPTION
**Pull request by adriaanm**
Follow up for #5294 based on @retronym's feedback.

I'm going to defer teaching the optimizer about stack allocating the lazy holders: https://github.com/scala/scala-dev/issues/215

### CODE REVIEW

**Comment by lrytz** [comment]
LGTM!

**Comment by retronym** [comment]
LGTM

File: src/library/scala/runtime/LazyRef.scala [hunk 1]
```diff
@@ -1,52 +1,157 @@
...
+
 package scala.runtime
 
-/* Classes used as holders for local lazy vals */
+/** Classes used as holders for lazy vals defined in methods. */
 
 class LazyRef[T] {
-  var value: T = _
-  @volatile var initialized: Boolean = false
+  @volatile private[this] var _initialized: Boolean = _
```
Comment (viktorklang) [inline src/library/scala/runtime/LazyRef.scala:14]: Why is this @volatile btw?

Comment (adriaanm) [reply src/library/scala/runtime/LazyRef.scala:14]: Quoting @retronym:
> In double checked locking, the write to the value is in a synchronized block
> the fast path (already initialized) just needs to volatile read on the initialized field. if that is set, we know the value was safely published

Comment (viktorklang) [reply src/library/scala/runtime/LazyRef.scala:14]: Apologies, I guess what threw me of course is that this class (or rather these classes) are public, and non-final, classes which cannot be used properly without sprinkling of "magic"/macros. Are these intended to be possible to use *directly* by users?

**Comment by adriaanm** [comment]
True, the runtime package is not meant for public consumption. Good point
that they should likely be final. They have to be public though, as they
are referenced from compiler-generated code.

File: src/library/scala/runtime/LazyRef.scala (hunk 1 above)
Comment (adriaanm) [reply src/library/scala/runtime/LazyRef.scala:14]: True, the runtime package is not meant for public consumption. Good point
that they should likely be final. They have to be public though, as they
are referenced from compiler-generated code.

https://github.com/scala/scala-dev/issues/339

**Pull request merged**